### DISCUSSION TITLE
Custom Data Types

### DISCUSSION THREAD

**Discussion by MikeBakerCZ** [taocpp/taopq#10] [Q&A] [answered]
Hi,
can you please put some examples how to implement custom data types (like datetime, timestamp, array etc) ?

Thanks!

**Comment by tkozybski** [comment]
Hello.
I've been playing around with taopq and this works for me (for retrieving). Needs to specialize tao::pq::parameter_traits to make it work with exec parameters:

```
struct Employee {
    int Id;
    std::string Name;
    int Age;
    std::string Address;
    double Salary;

    Employee(int id, std::string name, int age, std::string address, double salary)
        : Id{ id }
        , Name{ std::move(name) }
        , Age{ age }
        , Address{ std::move(name) }
        , Salary{ salary }

    {
    }
};

namespace tao::pq {
template <>
struct result_traits<Employee> {
    [[nodiscard]] static Employee from(row value)
    {

        return {
            value.get<int>(0),
            value.get<std::string>(1),
            value.get<int>(2),
            value.get<std::string>(3),
            value.get<double>(4),
        };
    }

    static constexpr size_t size = 5;
};
}
```

This code works now
```
        auto res = conn->execute("SELECT * FROM COMPANY ORDER BY ID").vector<Employee>();
        for (auto const& entry : res) {
            std::cout
                << entry.Id
                << " : " << entry.Name
                << " : " << entry.Age
                << " : " << entry.Address
                << " : " << entry.Salary
                << '\n';
        }
```

**Comment by d-frey** [comment]
Sorry I haven't found the time for providing an answer to this question. In fact, I think I might need to think about a few things first, as the parameter traits are currently not convenient and safe enough, so I will think about some improvements before I can finally add an example and some documentation.

**Comment by omerbn** [comment]
> Sorry I haven't found the time for providing an answer to this question. In fact, I think I might need to think about a few things first, as the parameter traits are currently not convenient and safe enough, so I will think about some improvements before I can finally add an example and some documentation.


Any news?

**Comment by d-frey** [comment, marked as answer]
Yes, I recently started to refactor some parameter-traits (now going for leaving a choice between text-based and binary-based parameter traits). It's not yet finished, but check out the example at https://github.com/taocpp/taopq/blob/e1e5cc016b8f7f1bb0d4e61f8f79a6bc448e5cad/src/test/pq/traits.cpp#L10-L42

`user` shows how to use a member method to allow a type to be used as a parameter for taopq, while `user2` shows how to do so with an external function when you can't modify a type.

I still need to figure out a way how to select the traits in a better way as well as combine different traits properly. Also, I probably need to improve the result traits as well. Still a lot to do and not enough time. But maybe the current progress already helps?

**Comment by omerbn** [comment]
When we'll I be able to use this ? :)

**Comment by d-frey** [comment]
You can use it right now. I think `to_taopq_param` (member- or free-function) will be the "convenience customization point" for the future, only more sophisticated use cases will require actual specializations of the parameter traits class. There will be other improvements (hopefully, when I have more time), but I'll try not to break compatibility unnecessarily.
